# PyCity Schools Analysis

My analysis

In [64]:
import pandas as pd
from pathlib import Path


school_data_path = Path("Resources/schools_complete.csv")
student_data_path = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_path)
student_data = pd.read_csv(student_data_path)

# Combine the data into a single dataset.  
school_data_total = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_total.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

In [65]:
 # Calculate the total number of unique schools
school_count = len(school_data_total["school_name"].unique())
print(school_count)

15


In [66]:
# Calculate the total number of students
student_count = len(school_data_total["student_name"])
print(student_count)

39170


In [67]:
# Calculate the total budget
total_budget = school_data["budget"].sum()
print(total_budget)

24649428


In [68]:
# Calculate the average (mean) math score
average_math_score = school_data_total["math_score"].mean()
print(average_math_score)

78.98537145774827


In [69]:
# Calculate the average (mean) reading score
average_reading_score = school_data_total["reading_score"].mean()
print(average_reading_score)

81.87784018381414


In [70]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
passing_math_count = school_data_total[(school_data_total["math_score"] >= 70)].count()["student_name"]
passing_math_percentage = passing_math_count / float(student_count) * 100
print(passing_math_percentage)

74.9808526933878


In [71]:
# Calculate the percentage of students who passed reading (hint: look at how the math percentage was calculated)  
passing_reading_count = school_data_total[(school_data_total["reading_score"] >= 70)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(student_count) * 100
print(passing_reading_percentage)

85.80546336482001


In [72]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = school_data_total[
    (school_data_total["math_score"] >= 70) & (school_data_total["reading_score"] >= 70)
].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(student_count) * 100
print(overall_passing_rate)

65.17232575950983


In [73]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
data = {"Total Schools":[school_count],"Total Students":[student_count],
        "Total Budget":[total_budget],"Average Math Score":[average_math_score],
        "Average Reading Score":[average_reading_score],"% Passing Math":[passing_math_percentage],
        "% Passing Reading":[passing_reading_percentage],"% Overall Passing":[overall_passing_rate]}

district_summary = pd.DataFrame(data)

# Formatting
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)

# Display the DataFrame
district_summary.head()

Total Schools Total Students    Total Budget  Average Math Score  \
0             15         39,170  $24,649,428.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

## School Summary

In [74]:
# show school names
df = school_data.groupby("school_name").count().reset_index()
df[["school_name"]].head()

school_name
0    Bailey High School
1   Cabrera High School
2  Figueroa High School
3      Ford High School
4   Griffin High School

In [75]:
#show school types
grp = school_data_total.groupby(["school_name"])
school_types = pd.Series(grp.first()["type"])
school_types.head().to_frame("School Type")

School Type
school_name                     
Bailey High School      District
Cabrera High School      Charter
Figueroa High School    District
Ford High School        District
Griffin High School      Charter

In [76]:
#total student count per school
per_school_counts = school_data_total.groupby("school_name").size()
per_school_counts.head().map("{:,}".format).to_frame("Student_Count")

Student_Count
school_name                       
Bailey High School           4,976
Cabrera High School          1,858
Figueroa High School         2,949
Ford High School             2,739
Griffin High School          1,468

In [77]:
# total school budget
grp = school_data_total.groupby("school_name")
per_school_budget = pd.Series(grp.first()[["budget"]].squeeze())
per_school_budget.head().map("${:,.2f}".format).to_frame("School Budget")

School Budget
school_name                        
Bailey High School    $3,124,928.00
Cabrera High School   $1,081,356.00
Figueroa High School  $1,884,411.00
Ford High School      $1,763,916.00
Griffin High School     $917,500.00

In [78]:
# per student budget
df = school_data_total.groupby("school_name")
per_student_budget = (df["budget"].mean() / df["student_name"].count())
per_student_budget.head().map("${:,.2f}".format).to_frame("Per Student Budget")

Per Student Budget
school_name                            
Bailey High School              $628.00
Cabrera High School             $582.00
Figueroa High School            $639.00
Ford High School                $644.00
Griffin High School             $625.00

In [79]:
# average math score
Average_Math = school_data_total.groupby("school_name")["math_score"].mean()
Average_Math.head().map("{:.2f}".format).to_frame("Average Math Score")

Average Math Score
school_name                            
Bailey High School                77.05
Cabrera High School               83.06
Figueroa High School              76.71
Ford High School                  77.10
Griffin High School               83.35

In [80]:
# average reading score
Average_Reading = school_data_total.groupby("school_name")["reading_score"].mean()
Average_Reading.head().map("{:.2f}".format).to_frame("Average Reading Score")

Average Reading Score
school_name                               
Bailey High School                   81.03
Cabrera High School                  83.98
Figueroa High School                 81.16
Ford High School                     80.75
Griffin High School                  83.82

In [96]:
# number of students who passed math and reading
df = school_data_total
passed_math_all = df.loc[(df["math_score"]) >=70]
passed_reading_all = df.loc[(df["reading_score"]) >=70]
passed_overall_all = df.loc[(df["math_score"] >=70) & (df["reading_score"]>=70)]

In [82]:
#  percentage of students who passed math per school
passed_math = passed_math_all.groupby("school_name").count()["student_name"] / per_school_counts * 100
passed_math.head().map("{:.2f}%".format).to_frame("Passed Math")
# grp_school_data = school_data_total.groupby("school_name")
# grp_students = grp_school_data["student_name"].count()s
# passed_math = school_data_total.query("math_score >= 70")["student_name"].count() / grp_students
#per_passed_math 

Passed Math
school_name                     
Bailey High School        66.68%
Cabrera High School       94.13%
Figueroa High School      65.99%
Ford High School          68.31%
Griffin High School       93.39%

In [83]:
# # percentage of students who passed reading
# grp_school_data = school_data_total.groupby("school_name")
# grp_students = grp_school_data["student_name"].count()
# passed_reading = school_data_total.query("reading_score >= 70")["student_name"].count() / grp_students
# passed_reading.head().map("{:.2f}%".format).to_frame("Passed Reading")
passed_reading = passed_reading_all.groupby("school_name").count()["student_name"] / per_school_counts * 100
passed_reading.head().map("{:.2f}%".format).to_frame("Passed Reading")


Passed Reading
school_name                        
Bailey High School           81.93%
Cabrera High School          97.04%
Figueroa High School         80.74%
Ford High School             79.30%
Griffin High School          97.14%

In [84]:
# percentage of overall students who passed math and reading
passed_overall = ((Average_Math + Average_Reading)/2)
passed_overall.head().map("{:.2f}%".format).to_frame("Passed Overall")

Passed Overall
school_name                        
Bailey High School           79.04%
Cabrera High School          83.52%
Figueroa High School         78.93%
Ford High School             78.92%
Griffin High School          83.58%

In [85]:
# create a school summary dataframe
dfdict = {"School Type":school_types,
          "Total Students":per_school_counts,
          "Total School Budget":per_school_budget,
          "Per Student Budget":per_student_budget,
          "Average Math Score":Average_Math,
          "Average Reading Score":Average_Reading,
          "% Passing Math":passed_math,
          "% Passing Reading":passed_reading,
          "% Overall Passing":passed_overall
          }

per_school_summary = pd.DataFrame(dfdict)
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)

per_school_summary.index_name = None
per_school_summary.head()

School Type  Total Students Total School Budget  \
school_name                                                            
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
school_name                                                 
Bailey High School            81.933280          79.041198  
Cabrera High School           97.039828          83.518837  
Figueroa High School          80.739234          78.934893  
Ford High School              79.299014          78.924425  
Griffin High School           97.138965          83.584128

## Highest-Performing Schools(by % Overall Passing)

In [86]:
#highest performing schools
top_schools = per_school_summary.sort_values(["% Overall Passing"], ascending=False)
top_schools.head()

School Type  Total Students Total School Budget  \
school_name                                                          
Pena High School       Charter             962         $585,858.00   
Wright High School     Charter            1800       $1,049,400.00   
Holden High School     Charter             427         $248,087.00   
Thomas High School     Charter            1635       $1,043,130.00   
Wilson High School     Charter            2283       $1,319,574.00   

                   Per Student Budget  Average Math Score  \
school_name                                                 
Pena High School              $609.00           83.839917   
Wright High School            $583.00           83.682222   
Holden High School            $581.00           83.803279   
Thomas High School            $638.00           83.418349   
Wilson High School            $578.00           83.274201   

                    Average Reading Score  % Passing Math  % Passing Reading  \
school_name                                                                    
Pena High School                84.044699       94.594595          95.945946   
Wright High School              83.955000       93.333333          96.611111   
Holden High School              83.814988       92.505855          96.252927   
Thomas High School              83.848930       93.272171          97.308869   
Wilson High School              83.989488       93.867718          96.539641   

                    % Overall Passing  
school_name                            
Pena High School            83.942308  
Wright High School          83.818611  
Holden High School          83.809133  
Thomas High School          83.633639  
Wilson High School          83.631844

## Lowest-Peforming Schools(by % Overall Passing)

In [87]:
# lowest performing schools
top_schools = per_school_summary.sort_values(["% Overall Passing"], ascending=True)
top_schools.head()

School Type  Total Students Total School Budget  \
school_name                                                             
Rodriguez High School    District            3999       $2,547,363.00   
Huang High School        District            2917       $1,910,635.00   
Ford High School         District            2739       $1,763,916.00   
Figueroa High School     District            2949       $1,884,411.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Rodriguez High School            $637.00           76.842711   
Huang High School                $655.00           76.629414   
Ford High School                 $644.00           77.102592   
Figueroa High School             $639.00           76.711767   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Rodriguez High School              80.744686       66.366592   
Huang High School                  81.182722       65.683922   
Ford High School                   80.746258       68.309602   
Figueroa High School               81.158020       65.988471   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School          80.220055          78.793698  
Huang High School              81.316421          78.906068  
Ford High School               79.299014          78.924425  
Figueroa High School           80.739234          78.934893  
Johnson High School            81.222432          79.019429

# Math Scores by Grade

In [88]:
# average students math scores by grade per school
df = school_data_total
grade_nine = df[(df["grade"] == "9th")]
grade_ten = df[(df["grade"] == "10th")]
grade_eleven = df[(df["grade"] == "11th")]
grade_twelve = df[(df["grade"] == "12th")]

scoretype = "math_score"

scores_nine = grade_nine.groupby("school_name")[scoretype].mean()
scores_ten = grade_ten.groupby("school_name")[scoretype].mean()
scores_eleven = grade_eleven.groupby("school_name")[scoretype].mean()
scores_twelve = grade_twelve.groupby("school_name")[scoretype].mean()

math_scores_by_grade = pd.DataFrame({"9th":scores_nine,"10th":scores_ten,
                                     "11th":scores_eleven,"12th":scores_twelve})

math_scores_by_grade.index.name = None
math_scores_by_grade.head()

9th       10th       11th       12th
Bailey High School    77.083676  76.996772  77.515588  76.492218
Cabrera High School   83.094697  83.154506  82.765560  83.277487
Figueroa High School  76.403037  76.539974  76.884344  77.151369
Ford High School      77.361345  77.672316  76.918058  76.179963
Griffin High School   82.044010  84.229064  83.842105  83.356164

# Reading Scores by Grade

In [89]:
# average students reading scores by grade per school
df = school_data_total
grade_nine = df[(df["grade"] == "9th")]
grade_ten = df[(df["grade"] == "10th")]
grade_eleven = df[(df["grade"] == "11th")]
grade_twelve = df[(df["grade"] == "12th")]

scoretype = "reading_score"

scores_nine = grade_nine.groupby("school_name")[scoretype].mean()
scores_ten = grade_ten.groupby("school_name")[scoretype].mean()
scores_eleven = grade_eleven.groupby("school_name")[scoretype].mean()
scores_twelve = grade_twelve.groupby("school_name")[scoretype].mean()

reading_scores_by_grade = pd.DataFrame({"9th":scores_nine,"10th":scores_ten,
                                     "11th":scores_eleven,"12th":scores_twelve})

reading_scores_by_grade.index.name = None
reading_scores_by_grade.head()

9th       10th       11th       12th
Bailey High School    81.303155  80.907183  80.945643  80.912451
Cabrera High School   83.676136  84.253219  83.788382  84.287958
Figueroa High School  81.198598  81.408912  80.640339  81.384863
Ford High School      80.632653  81.262712  80.403642  80.662338
Griffin High School   83.369193  83.706897  84.288089  84.013699

# Scores by School Spending

In [90]:
# create the bins
spending_bins = [0, 585, 630, 645, 680]
spending_labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [91]:
# copy per_school_summary
school_summary_df = per_school_summary.copy()

In [92]:
# Use 'pd_cut' to categorize spending based on the bins
school_summary_df["Spending Ranges (Per Student)"] = pd.cut(per_student_budget, spending_bins, labels=spending_labels)

spending_math_scores = school_summary_df.groupby(["Spending Ranges (Per Student)"])["Average Math Score"].mean()
spending_reading_scores = school_summary_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
spending_passing_math = school_summary_df.groupby(["Spending Ranges (Per Student)"])["% Passing Math"].mean()
spending_passing_reading = school_summary_df.groupby(["Spending Ranges (Per Student)"])["% Passing Reading"].mean()
overall_passing_spending = school_summary_df.groupby(["Spending Ranges (Per Student)"])["% Overall Passing"].mean()

spending_summary = pd.DataFrame({"Average Math Score":spending_math_scores,
                                 "Average Reading Score":spending_reading_scores,
                                 "% Passing Math":spending_passing_math,
                                 "% Passing Reading":spending_passing_reading,
                                 "% Overall Passing":overall_passing_spending})
spending_summary

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-630                                81.899826              83.155286   
$630-645                                78.518855              81.624473   
$645-680                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-630                            87.133538          92.718205   
$630-645                            73.484209          84.391793   
$645-680                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                  83.694607  
$585-630                               82.527556  
$630-645                               80.071664  
$645-680                               79.012526

# Score by School Size

In [93]:
# create the bins
size_bins = [0, 1000, 2000, 5000]
size_labels = ["Small (<1000)", "Medium(1000-2000)", "Large (2000-5000)"]

In [94]:
# Use 'pd_cut' to categorize spending based on the bins
per_school_summary["Total Students"] = pd.cut(per_school_counts, size_bins, labels=size_labels)

size_math_scores = per_school_summary.groupby(["Total Students"])["Average Math Score"].mean()
size_reading_scores = per_school_summary.groupby(["Total Students"])["Average Reading Score"].mean()
size_passing_math = per_school_summary.groupby(["Total Students"])["% Passing Math"].mean()
size_passing_reading = per_school_summary.groupby(["Total Students"])["% Passing Reading"].mean()
size_overall_passing = per_school_summary.groupby(["Total Students"])["% Overall Passing"].mean()

size_summary = pd.DataFrame({"Average Math Score":size_math_scores,
                             "Average Reading Score": size_reading_scores,
                             "% Passing Math":size_passing_math,
                             "% Passing Reading":size_passing_reading,
                             "% Overall Passing":size_overall_passing})

size_summary

Average Math Score  Average Reading Score  % Passing Math  \
Total Students                                                                 
Small (<1000)               83.821598              83.929843       93.550225   
Medium(1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)           77.746417              81.344493       69.963361   

                   % Passing Reading  % Overall Passing  
Total Students                                           
Small (<1000)              96.099437          83.875721  
Medium(1000-2000)          96.790680          83.619561  
Large (2000-5000)          82.766634          79.545455

# Scores by School Type

In [95]:
# Group the per_school_summary DataFrame by "School Type" and average the results
average_math_score_by_type = per_school_summary.groupby(["School Type"])["Average Math Score"].mean()
average_reading_score_by_type = per_school_summary.groupby(["School Type"])["Average Reading Score"].mean()
average_percent_passing_math_by_type = per_school_summary.groupby(["School Type"])["% Passing Math"].mean()
average_percent_passing_reading_by_type = per_school_summary.groupby(["School Type"])["% Passing Reading"].mean()
average_percent_overall_passing_by_type = per_school_summary.groupby(["School Type"])["% Overall Passing"].mean()

type_summary = pd.DataFrame({"Average Math Score":average_math_score_by_type,
                             "Average Reading Score":average_reading_score_by_type,
                             "% Passing Math":average_percent_passing_math_by_type,
                             "% Passing Reading":average_percent_passing_reading_by_type,
                             "% Overall Passing":average_percent_overall_passing_by_type})

type_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          83.685136  
District             80.799062          78.961685

## Report: My Analysis

Generally, charter schools perform better than districts, as shown by the data analysis the top and bottom 5 schools.
Across grades 9-12, the average reading scores seem to be slightly higher than the math scores.
Students demosntrate consistent performances in math and reading, as shown by Bailey High School.
Bailey High School's 10th grade students had an average math score of 76.99 and an average score of 80.90 for reading.
Ford High School's percentage of students who passed math was 68.31% and for reading it was 79.30%.